### PR TITLE
Add/Remove multiple entries with Scoreboard Team

### PR DESCRIPTION
**Pull request by Cryptite**
It is a common pattern to fill a team up with several entries. Unfortunately, each single entry addition sends a ScoreboardTeam packet out which can be unnecessarily chatty. This patch just adds the option to set a collection of entries on a scoreboard team which can get bundled into a single packet.

Likely not a significant optimization, but hey, every little bit helps.

### CODE REVIEW

File: patches/api/0333-Multiple-Entries-with-Scoreboards.patch [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+
+diff --git a/src/main/java/org/bukkit/scoreboard/Team.java b/src/main/java/org/bukkit/scoreboard/Team.java
+index d5b39fb4fc16a342b5661e08df1506858168d20d..9ec00b045fe85701ce29c0ad10eb9921733107b3 100644
+--- a/src/main/java/org/bukkit/scoreboard/Team.java
++++ b/src/main/java/org/bukkit/scoreboard/Team.java
+@@ -1,5 +1,6 @@
+ package org.bukkit.scoreboard;
+ 
++import java.util.Collection; // Paper
```
Comment (Machine-Maker) [inline patches/api/0333-Multiple-Entries-with-Scoreboards.patch]: Use fully qualified name in the 2 places below

File: patches/server/0807-Multiple-Entries-with-Scoreboards.patch [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+         this.method = buf.readByte();
+diff --git a/src/main/java/net/minecraft/server/ServerScoreboard.java b/src/main/java/net/minecraft/server/ServerScoreboard.java
+index 130a928f156961bae9ca184b3ca31004dbba1012..5d42967b1d9cbceb004566c0c503df21d1bb5c51 100644
+--- a/src/main/java/net/minecraft/server/ServerScoreboard.java
++++ b/src/main/java/net/minecraft/server/ServerScoreboard.java
+@@ -3,6 +3,7 @@ package net.minecraft.server;
+ import com.google.common.collect.Lists;
+ import com.google.common.collect.Sets;
+ import java.util.Iterator;
++import java.util.Collection; // Paper
```
Comment (Machine-Maker) [inline patches/server/0807-Multiple-Entries-with-Scoreboards.patch]: Same as above, use FQN below

File: patches/server/0807-Multiple-Entries-with-Scoreboards.patch [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+     }
+diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
+index 2b87a652798cb632fe76bf20e9e7f8cb8bfb3b7b..f097e9fbdac1a572c12cc0299774455c75e373e9 100644
+--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
++++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
+@@ -1,6 +1,7 @@
+ package org.bukkit.craftbukkit.scoreboard;
+ 
+ import com.google.common.collect.ImmutableSet;
++import java.util.Collection; // Paper
```
Comment (Machine-Maker) [inline patches/server/0807-Multiple-Entries-with-Scoreboards.patch]: same here

**Review by Machine-Maker — CHANGES_REQUESTED** [review]
Why can't you use the existing super.addPlayerToTeam method to add each player instead of creating your own for adding/removing a collection of entries?

**Comment by Cryptite** [comment]
I suppose the only reason (from a convo in the previous PR), if I do that and one of the entries is bad, the whole method throws with whatever was added so far in the team. The only real difference is the collection version of the method warns and moves on, rather than throws. 

I'm unsure what's the better option here to be honest.

**Comment by Machine-Maker** [comment]
I think it should throw an exception either way, right now you are just logging errors. The plugin is doing something wrong if it tries to add an invalid entry, and that shouldn't be swept under the rug with a simple warning message.

**Comment by Cryptite** [comment]
Reduced diffs. The reason I can't use the existing super is because it is the reason that a packet-per-entry is fired. There's not really a way to nicely call the existing code that doesn't get overridden by the methods on `ServerScoreboard`

File: patches/server/0807-Multiple-Entries-with-Scoreboards.patch [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+         super.onObjectiveAdded(objective);
+diff --git a/src/main/java/net/minecraft/world/scores/Scoreboard.java b/src/main/java/net/minecraft/world/scores/Scoreboard.java
+index 3e75ea4d5a6c83ca570b29e3c1a5d51fb132379a..48164cd490c8105a733752df5ab45fa05ad4470b 100644
+--- a/src/main/java/net/minecraft/world/scores/Scoreboard.java
++++ b/src/main/java/net/minecraft/world/scores/Scoreboard.java
+@@ -2,12 +2,14 @@ package net.minecraft.world.scores;
+ 
+ import com.google.common.collect.Lists;
+ import com.google.common.collect.Maps;
++
```
Comment (Machine-Maker) [inline patches/server/0807-Multiple-Entries-with-Scoreboards.patch]: remote

File: patches/server/0807-Multiple-Entries-with-Scoreboards.patch [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+ import java.util.Collection;
+ import java.util.Collections;
+ import java.util.List;
+ import java.util.Map;
+ import java.util.function.Consumer;
+ import javax.annotation.Nullable;
++
```
Comment (Machine-Maker) [inline patches/server/0807-Multiple-Entries-with-Scoreboards.patch]: this too

File: patches/server/0807-Multiple-Entries-with-Scoreboards.patch [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
++        for (Map<Objective, Score> map : this.playerScores.values()) {
+             map.remove(objective);
+         }
+ 
+@@ -204,7 +206,7 @@ public class Scoreboard {
+     public void removePlayerTeam(PlayerTeam team) {
+         this.teamsByName.remove(team.getName());
+ 
+-        for(String string : team.getPlayers()) {
++        for (String string : team.getPlayers()) {
```
Comment (Machine-Maker) [inline patches/server/0807-Multiple-Entries-with-Scoreboards.patch]: think you accidentally changed formatting here

File: patches/server/0807-Multiple-Entries-with-Scoreboards.patch [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
++    }
++    // Paper end
++
+     @Override
+     public void onObjectiveAdded(Objective objective) {
+         super.onObjectiveAdded(objective);
+diff --git a/src/main/java/net/minecraft/world/scores/Scoreboard.java b/src/main/java/net/minecraft/world/scores/Scoreboard.java
+index 3e75ea4d5a6c83ca570b29e3c1a5d51fb132379a..48164cd490c8105a733752df5ab45fa05ad4470b 100644
+--- a/src/main/java/net/minecraft/world/scores/Scoreboard.java
++++ b/src/main/java/net/minecraft/world/scores/Scoreboard.java
```
Comment (lynxplay) [inline patches/server/0807-Multiple-Entries-with-Scoreboards.patch:78]: Seems like you reformatted this file. Please remove these reformatting changes.

**Review by Machine-Maker — CHANGES_REQUESTED** [review]
I think you accidentally increased diff with a bunch of whitespace changes in the server patch.

**Comment by Machine-Maker** [comment]
> Reduced diffs. The reason I can't use the existing super is because it is the reason that a packet-per-entry is fired. There's not really a way to nicely call the existing code that doesn't get overridden by the methods on `ServerScoreboard`

You can call them with `super.addPlayerToTeam()` right? inside the loop iterating over the new entries to add?
That's gonna still send 1 packet per entry removed from the team, but your existing code still already did that.

**Comment by Cryptite** [comment]
This was inadvertently deleted and became https://github.com/PaperMC/Paper/pull/6640

**Pull request closed without merging**